File: Chapter 02/PrototypeDesignPatternExample.py
import copy

class PrototypeExample:
    def clone(self):
        pass
class ConcretePrototypeExample(PrototypeExample):
    def __init__(self, name):
        self.name = name
    def clone(self):
        return copy.deepcopy(self)
    def get_name(self):
        return self.name

if __name__ == "__main__":
    original = ConcretePrototypeExample("Original Prototype")
    copy = original.clone()

    print("Original:", original.get_name())
    print("Copy:", copy.get_name())

File: Chapter 02/PrototypeDesignPatternExample.java
interface PrototypeExample {
    PrototypeExample clone();
}


class ConcretePrototypeExample implements PrototypeExample {
    private String name;
    public ConcretePrototypeExample(String name) {
        this.name = name;
    }
    public PrototypeExample clone() {
        return new ConcretePrototypeExample(this.name);
    }
    public String getName() {
        return name;
    }
}

public class PrototypeDesignPatternExample {
    public static void main(String[] args) {
        PrototypeExample original = new ConcretePrototypeExample("Original Prototype");
        PrototypeExample copy = original.clone();

        System.out.println("Original: " + original.getName());
        System.out.println("Copy: " + copy.getName());
    }
}
